### PR TITLE
ENH: Adding support to the range keyword for estimation of the optimal number of bins and associated tests

### PR DESCRIPTION
**Pull request by madphysicist**
This PR is a quick fixup of #6288 made on behalf of @nayyarv. I was unable to make a PR directly back into Varun's repo for him to update his branch.

### CODE REVIEW

**Comment by nayyarv** [comment]
I'm happy with it, thanks @madphysicist.

 @seberg, @njsmith

**Comment by nayyarv** [comment]
I also like the idea of raising a `RunTimeWarning` instead of a `TypeError` when weighted data is passed in.

**Comment by njsmith** [comment]
Looks good to me, thanks @nayyarv and @madphysicist!

I feel strongly that we should raise the error rather than print a warning and return possibly nonsensical results. If nothing else, once we start returning nonsensical results we may get stuck doing that forever due to backcompat issues, while a error is something we can easily fix once we decide what the right answer is.

**Comment by njsmith** [comment]
@charris I think this should get a backport to 1.11, in such case it doesn't need a release note because it will be part of the initial implementation of automatic bin selection (which should already have a release note). I'll leave you to make the final decision and hit merge here (or not), though?

**Comment by charris** [comment]
@njsmith A backport would be fine with me. So if you want to merge this, remove the release note tag and go ahead.

File: numpy/lib/function_base.py [hunk 1]
```diff
@@ -84,15 +84,34 @@ def _hist_optim_numbins_estimator(a, estimator):
...
 
+    if data_weights is not None:
+        raise TypeError("Automated estimation of the number of "
+                        "bins is not supported for weighted data")
+
+    if data_range is not None:
+        mn, mx = data_range
+        keep = (a >= mn)
+        keep &= (a <= mx)
+        if not np.logical_and.reduce(keep):
```
Comment (shoyer) [inline numpy/lib/function_base.py]: use `np.all(keep)` instead

Comment (madphysicist) [inline numpy/lib/function_base.py]: I agree that maintainability and legibility outweighs a fractional improvement in speed. However, @nayyarv made a pretty good argument for keeping it this way. Could I have one more person weigh in to break the tie?

**Comment by madphysicist** [comment]
I have been doing some thinking about how to do weighted partitioning for functions like median and percentile. Looking at the existing C code, it should be a reasonable amount of work to add a `weights` parameter to all those functions as long as the weights are restricted to reals. I don't think that a separate module is a good idea since weights are meaningless for many of the functions in `function_base`, just an extra parameter for the relevant ones will be fine. Is there interest in me proceeding with this?

File: numpy/lib/function_base.py (hunk 1 above)
Comment (njsmith) [inline numpy/lib/function_base.py]: Previous discussion with @nayyarv's comment was here: https://github.com/numpy/numpy/pull/6288#discussion_r38879392

I.e., this file uses `and.reduce` elsewhere for this same computation. I'm not too bothered either way, using `add.reduce` consistently isn't really worse than using `all` in some places and `add.reduce` in others, and I'd like to get this moving for 1.11, so @shoyer -- I'm going to merge this and we can figure out whether we care about `add.reduce` versus `all` as a separate matter? Feel free to squawk if you think this is a grave mistake :-)

**Comment by njsmith** [comment]
@charris: Okay, merging

**Comment by njsmith** [comment]
@madphysicist: that sounds like useful functionality in general, sure, but this probably isn't the place to discuss it :-). The mailing list or a new issue might be a good place if you want to get some feedback on the general idea...

File: numpy/lib/function_base.py (hunk 1 above)
Comment (shoyer) [inline numpy/lib/function_base.py]: Oops... I forgot I already made this comment! 

This is fine :)

Comment (njsmith) [inline numpy/lib/function_base.py]: I guess this was an inadverdant double-blind test of your review consistency. You did [better than NIPS](http://blog.mrtz.org/2014/12/15/the-nips-experiment.html) ;-)

Comment (charris) [inline numpy/lib/function_base.py]: There are some advantages to `np.all`
```
    arr = asanyarray(a)
    kwargs = {}
    if keepdims is not np._NoValue:
        kwargs['keepdims'] = keepdims
    return arr.all(axis=axis, out=out, **kwargs)
```
I don't know if they apply here, but I do know that when I did the initial nanfunctions using the higher level `np.sum` and such was urged upon me and did indeed improve behavior with subclasses and such.

**Pull request merged**